### PR TITLE
Fix hang after calling `SaeTrainer.save()` w/ DDP

### PR DESCRIPTION
**Pull request by vgel**
While trying to train with DDP on an 8xA100 node with default arguments...

```bash
torchrun --nproc_per_node gpu -m sae EleutherAI/pythia-160m togethercomputer/RedPajama-Data-1T-Sample --attn_implementation=eager
```

...the trainer hangs after saving--the layers are successfully saved, but afterwards the trainer doesn't make progress and stays stuck on iteration 999.

Running with `TORCH_DISTRIBUTED_DEBUG=DETAIL`, when a normal run would've hung, the trainer instead crashes with a process desync exception, saying rank 0 is in a barrier while the others have moved on to reduce. (Unfortunately I didn't save the exact exception, if it'd be useful I can undo my fix and get it.)

I believe the error is in `trainer.py:save`:

https://github.com/EleutherAI/sae/blob/2cc21c8ad11140209c725de68219ccfc5cea2c18/sae/trainer.py#L296-L309

The first `if` causes the other processes to leave the function immediately, so only rank 0 hits the barrier, causing the desync. After changing `save` to remove the early return, so that all processes reach `dist.barrier()`, training now passes the savepoint without issue.

### CODE REVIEW

**Review by norabelrose — APPROVED** [review]
LGTM thanks

File: sae/trainer.py [hunk 1]
```diff
@@ -294,15 +294,15 @@ def scatter_hiddens(self, hidden_list: list[Tensor]) -> list[Tensor]:
 
     def save(self):
         """Save the SAEs to disk."""
-        if (dist.is_initialized() and dist.get_rank() != 0) and not self.cfg.distribute_layers:
-            return
 
-        for i, sae in zip(self.cfg.layers, self.saes):
-            assert isinstance(sae, Sae)
-            print(f"Saving layer {i}")
+        if self.cfg.distribute_layers or dist.get_rank() == 0 or not dist.is_initialized():
```
Comment (vgel) [inline sae/trainer.py:298]: I think get_rank will throw if `not dist.is_initialized` so we should flip the cond:

```suggestion
        if self.cfg.distribute_layers or not dist.is_initialized() or dist.get_rank() == 0:
```

Comment (norabelrose) [reply sae/trainer.py:298]: I fixed this locally before seeing this lol

**Review by norabelrose — APPROVED** [review]
LGTM

**Pull request merged**